### ISSUE TITLE
Also make linux releases!

### ISSUE THREAD

**Issue by Altonss** [Audiveris/audiveris#534]
Hello,
I just discovered the project and am very interested! The project seems amazing and I want to help bringing the software on linux. I saw we can already build it from source and that's already great. But It would be even better provide builds, for example through AppImages or Flatpaks :)
If I find the time to help with this issue, I would be happy to contribute!

**Comment by hbitteur** [comment]
Sure, our small team would be happy to benefit from such contribution!

Personally I develop on Windows and my competency about Linux is close to zero. So I will make sure that Max, by far our most competent Linux guy,  is aware of your proposal.

I know nothing about AppImages and Flatpaks.
I think I remember that Oracle is to provide a general binary-packaging utility in some next Java release (perhaps based on jlink?). A track to explore.
We just have a Windows-specific installer today, a more general approach would be interesting of course.

**Comment by Altonss** [comment]
I made some progress to package Audiveris as a flatpak. For the moment it is using the build obtained using gradle and wraps this into a flatpak. That's not very clean, but at least it's a beginning. Should I share my recipe?
I could also try something with the jar files, but that's more complicated as I would need gradle to run these...Or am I wrong?

**Comment by Altonss** [comment]
Does Audiveris needs network access? Because I got some errors because of some lacking network access

**Comment by hbitteur** [comment]
Audiveris checks periodically for a potential new release.
This is the only automatic network access I can think of (and you can change the polling period which is weekly by default).
If there is no network, you should get just a warning like "Could not connect to Audiveris project".

Of course if you manually select, in the Help menu, items like Handbook, Wiki, Web Site, Check for Update, you will try to reach the information on the network.

**Comment by hbitteur** [comment]
Regarding flatpak, I can't help you. I will let Max answer your question if he can.

**Comment by mwilck** [comment]
I've created a flatpak recipe [here](https://github.com/mwilck/Audiveris-flatpak). I created a helper tool that allows downloading all Java dependencies and using them for a clean flatpak build using `grade --offline`. I have created 4 branches for packaging 5.2.4, 5.2.5, and the current `master` and `development` branches, respectively. These branches contain all required dependencies, so they can be used for a simple `flatpak-builder` invocation. The bootstrapping procedure for downloading the dependencies is explained in the [README](https://github.com/mwilck/Audiveris-flatpak/blob/development/README.md).
There are also two slim branches with no dependencies (`scripts` and `scripts-17` for openJDK11 and openJDK17, respectively).

Network access functions in this version. But opening a browser to read the docs currently doesn't work, this [seems to be a generic Java/flatpak issue](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk17/issues/1).

@Altonss, I just discovered this issue here after I was already done with my work. I hope you're not upset. Perhaps we can make a joint effort to get this ready for prime-time?

@hbitteur, would you agree to submitting these packages to flathub?

**Comment by Altonss** [comment]
Thanks a lot for your work @mwilck !
I already had tried to build the flatpak, and it kind of worked, but thanks for having made a clean version. Are all the functionnalities working?
It would be sooo cool to be able to release some day on Flathub :)

**Comment by mwilck** [comment]
This is still in it's early stages. I haven't checked if *everything* works (as I said, launching a browser does not). I've opened some sample music sheets and run OMR on it; I was quite satisfied with the results so far. I'd be very grateful if you could give my packages a try. You should be able to build them fairly easily (clone my repo, checkout branch `b5.2.5`, and run `flatpak-builder --repo /some/dir builddir org.audiveris.audivers.yml`).

**Comment by mwilck** [comment]
I've created a short [Wiki page](https://github.com/mwilck/Audiveris-flatpak/wiki/Home:-Branches-in-this-repository) explaining what the various branches are.

**Comment by mwilck** [comment]
The web browser problem is solved now due to help I got [in the openJDK17 issue](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk17/issues/1).

**Comment by mwilck** [comment]
@hbitteur: I'd appreciate a comment if you'd agree with submitting this to flathub, and if not, what you think I (we) should do beforehand. In particular, would you object me submitting this with the flatpak application ID `org.audiveris.audiveris`?

**Comment by hbitteur** [comment]
@mwilck 
I have no objection to your submitting.
And I'll be out for a couple of days so don't wait for other comments from my side, simply keep going as you intend.

**Comment by mwilck** [comment]
I haven't submitted to flathub yet, but I've reworked the build scripts such that no dependencies need to be stored locally. Everything is now downloaded by flatpak-builder before gradle builds offline.

**Comment by hbitteur** [comment]
Any progress on this topic?
We do need such Linux releases!

**Comment by mwilck** [comment]
Sorry I haven't been able to follow up here. I suppose my previous packaging work could be brought back to work with relatively little effort. I will try to find some time to work on it soon.

FYI, back in 2022 when I worked on this, I got frustrated because Audiveris would always crash in unpredictable ways in the text recognition stage, typically with references to freed memory, which AFAICT should never happen in Java. I am  certain that this wasn't caused by the flatpak, it also happened with native binaries, so maybe it was an issue with Tesseract, or possibly with my JRE (actually various different JREs I tried). I didn't want to submit something I knew was going to crash on non-trivial scan attempts. 

I guess I should have opened in issue here, but I thought that this can't be a generic issue but must be caused in some ways by my Linux environment. I didn't make any progress with the bug analysis though, and eventually gave up, especially as the project for which I had intended to use Audiveris was finished.

**Comment by mwilck** [comment]
So ... I had another look at my build environment for the flatpak, and to my surprise, I found that it still works smoothly on the lastest code. I have a  [branch for packaging Audiveris 5.3.1](https://github.com/mwilck/Audiveris-flatpak/tree/b5.3.1), and the `development` branch has all the meta data for packaging current Audiveris `development`. The other good news is that I couldn't reproduce the Tesseract crashes I'd previously seen.

So it seems that we're actually in quite decent shape. I'm sure there are still lots of issues (for example, the "About" page shows wrong versions, not idea why).

**Comment by mwilck** [comment]
@hbitteur:

The flatpak organization prefers that the author of an application submits the package to Flathub. Maybe you could consider doing the submission yourself? The procedure is described [here](https://docs.flathub.org/docs/for-app-authors/submission/). 

You need a fork and a clone of flathub itself, which must be populated with the necessary files to build the Audiveris package. I have created a sample flathub fork [here ("audiveris" branch)](https://github.com/mwilck/flathub/tree/audiveris). It was created from my [Audiveris-flatpak](https://github.com/mwilck/Audiveris-flatpak) repo, branch `b5.3.1`, using the script `flathub-copy.sh`.[^gen]

AFAICS you could use my flathub fork for a first submission attempt of Audiveris 5.3.1 without changes. However, please review it again, in particular the `org.audiveris.audiveris.metainfo.xml` file. If you find issues, or have any additional questions, contact me. Please also have a look at the [Flathub submission requirements](https://docs.flathub.org/docs/for-app-authors/requirements) page, licensing etc.

Note that Flathub's requirements wrt licensing aren't overly strict. In my understanding, if we can package a Windows binary under some open-source license and distribute it, we should be able to publish on Flathub, too. But IANAL.

If you can't do this submission yourself for whatever reason, let me know, and I'll try to figure out how I can do it in your name.

[^gen]: The purpose of the mwilck/Audiveris-flatpak code is to create a build manifest for building Audiveris offline, as required by flatpak. The code that will eventually be submitted to Flathub is *generated* by the code in the mwilck/Audiveris-flatpak repo.

**Comment by hbitteur** [comment]
@mwilck 

It's good news. Both for the Audiveris project and for all those who ask for easy installation on Linux.

We can add the fact that, thanks to [recent PR from Peter Greth](https://github.com/Audiveris/audiveris/pull/666), Audiveris was able to replace JPodRenderer with Apache PDFBox. JPodRenderer still had a dependency on JAI (Java Advanced Imaging) which was not fully open-source. The road is now clear.

Today I went through the Flatpak and Flathub documentations that you pointed out in your post. I also read the 15 files contained in your Flathub fork. Should I say that I was far from understanding every line?

So, I think you, Martin, are the person who should take the lead on this Flatpak/Flathub submission.
Indeed, I read that "*some applications are distributed on Flathub without the involvement of their developers*" and that Flathub "*would prefer that these applications be controlled by their authors*".
But that's not our case. I think all Audiveris contributors, starting with me, would appreciate this submission.

This Audiveris project is open in all directions, all contributions are welcome.
Baruch contributed with the Windows installer. This Linux distribution will be your contribution.

I will try to help you as much as I can.
For example, I think we should aim for a new version (5.3.2), due to some bug fixes I'd like to include.

We'll get into more detail in the coming days, but I wanted to tell you that you should go for it.
Thank you very much for your work.
/Hervé

P.S. I need to resurrect my VirtualBox installation to get my hands on the Linux environment again

**Comment by mwilck** [comment]
[Flathub submission created](https://github.com/flathub/flathub/pull/4633). Let's see how it goes. @hbitteur please have a look and let me know if anything is not as you would like it.

**Comment by mwilck** [comment]
@hbitteur, have you ever tried running Audiveris on other platforms than x86_64? The flatpak build fails for ARM because of missing tesseract libraries, and I wonder if it's worth the effort to try and make it build. For now I am just trying to disable the ARM build.

**Comment by hbitteur** [comment]
@mwilck 
No, I have never tried it myself.
For what it's worth, have a look at issue #564

**Comment by mwilck** [comment]
We made it! [Audiveris is now on flathub](https://flathub.org/apps/org.audiveris.audiveris). :rocket: 

I suppose this issue can be closed.

**Comment by mwilck** [comment]
@hbitteur, perhaps you want to add a note about this in your installation guide.

**Comment by hbitteur** [comment]
Sure, I need to do this!
Let me try to run the installation on my PC.

Thanks a lot!

**Comment by mwilck** [comment]
I haven't double-checked, but I suppose verification can be only done for a single account. I can do this, but I'd leave the decision to @hbitteur whether he wants to link the app to his own account instead, which would increase credibility.

**Comment by mwilck** [comment]
@Altonss, perhaps you could give the package a test?

**Comment by mwilck** [comment]
About the verification: the process requires me create a file with an UUID-like token in a file on the audiveris.org web page (https://audiveris.org/.well-known/org.flathub.VerifiedApps.txt). I obviously can't do that myself. So @hbitteur, I'll need your cooperation for creating that file.
Perhaps it's actually easier if you do this yourself. You can use github to identify to flathub.

**Comment by mwilck** [comment]
> Only issue is, by default `Export book` exports to `/home/username/.var/app/org.audiveris.audiveris/data/AudiverisLtd/audiveris/piece/piece.mvt1.mxl`, which isn't ideal for the end user I think (it would be better in documents for ex).

I agree that this path isn't ideal. I have kept changes to Audiveris's own source code to an absolute minimum. The only commit affecting it is 1f326d4.

Let me try to clarify my role here: I will will work on keeping the flatpak up to date and improving the code _for creating the flatpak package_. I don't feel responsible for other issues such as this one, even if they only occur on Linux / in the flatpak package. I will be happy to help with such issues, but I don't know the sources well enough to commit myself to fixing them.

In the concrete case here, I guess we need some indicator to figure out whether the app is being compiled for a) Linux and b) a flatpak environment, and set some compilation defaults accordingly. I'll try to figure something out, but it may take a while.

That said, I set the permissions of the flatpak such that it has access to `$HOME/Documents` and `$HOME/Pictures`. You should be able to use these folders.

**Comment by mwilck** [comment]
About the verification: I would need @hbitteur's [assistance](https://github.com/flathub/flathub/pull/4633#issuecomment-1785468450).

I think we should stop here. Adding more side issues to the "make linux release"  topic is confusing. Let's discuss the verification process here, all else should go into separate issues.

**Comment by hbitteur** [comment]
Hi guys,

I hadn't used VirtualBox since about 10 years...
So, I spent half a day fighting with that beast: downloading VirtualBox 7, installing Ubuntu unattended,
discovering that my (french) keyboard was stuck to Qwerty, that I could not open a terminal with CTRL+ALT+T...
Then a lot of googling to read somewhere that I should NOT use the default unattended way if I wanted to have the opportunity to correctly manage my Azerty keyboard.
So I reinstalled everything, this time keyboard was OK and I could open a terminal window.
I then installed flatpak, and was able to install audiveris from flathub like a charm.

Now add a couple of hours after dinner to succeed in sharing folders between my Windows host and the Ubuntu guest.
And voila, Audiveris now runs correctly.

![image](https://github.com/Audiveris/audiveris/assets/22053634/1f7c55f4-92b2-42a1-be19-d5b5ec3ea3c8)

As you can see, I have been able to store a resulting .mxl file.
The target is similar to what you obtained.

I'll see tomorrow if and how we can target other folders.
This is enough for today.

**Comment by hbitteur** [comment]
> Only issue is, by default `Export book` exports to `/home/username/.var/app/org.audiveris.audiveris/data/AudiverisLtd/audiveris/piece/piece.mvt1.mxl`, which isn't ideal for the end user I think (it would be better in documents for ex).

I agree with @mwilck: this has little to do with the main topic of this issue which is to provide a "Linux installer".
I'd suggest we open a separate issue or discussion if needed on this question of target folder.

See the specific [target folders](https://audiveris.github.io/audiveris/_pages/advanced/topics/) article and the more [general standard folders](https://audiveris.github.io/audiveris/_pages/folders/standard/) article in Audiveris handbook.

**Comment by hbitteur** [comment]
> I then installed flatpak, and was able to install audiveris from flathub like a charm.

This is truly a significant success! Thank you all for this long-awaited achievement!

Just like we did with Windows installer, we now need to automate this publishing mechanism.
Certainly by defining specific tasks in the Gradle script. We need to work with @mwilck on this.

**Comment by hbitteur** [comment]
I could not resist...
Here below is [Flathub web site](https://flathub.org/en) today.
Enjoy! 

![image](https://github.com/Audiveris/audiveris/assets/22053634/9188c162-b4cd-4b49-a275-2e853cb621ec)

**Comment by hbitteur** [comment]
> Just like we did with Windows installer, we now need to automate this publishing mechanism. Certainly by defining specific tasks in the Gradle script. We need to work with @mwilck on this.

Today, I tried to review and understand all the various pieces that Martin Wilck has worked upon to pave the way from Audiveris to Flathub.
My motivation is double:
- Be able to finally take the ownership, as strongly recommended by the Flathub team,
- Be able to automate the publishing process as much as possible.

I first went through all the files, trying to understand their purpose together with the Flathub documentation in the other hand, then I stumbled on Martin's [Audiveris-Flatpak README file](https://github.com/mwilck/Audiveris-flatpak/blob/development/README.md) which helped me a lot. I should have begun with it!

My understanding, please correct me if I am wrong,  is that a submission to Flathub is implemented as a "pull request" on a forked Flathub dedicated project.
And the PR must provide specifically populated files (a kind of manifest).
The question is how to write these files and, even more importantly, how to keep them up-to-date with Audiveris evolutions.
So that, for any new Audiveris release, we can easily publish the master source release on GitHub, the Windows installer, the documentation, etc ... and the Linux installer on Flathub.

Could we handle this (the process described in Martin's README file) via the `build.gradle` file hosted in Audiveris GitHub project? 
This is what we did for Windows installer, and for the schemas documentation as well (using a dedicated sub-project).
Problem: here we need to use tools like `flatpak-builder` which, to my knowledge, need to be run on Linux.

Could someone infirm or confirm this possibility?

**Comment by hbitteur** [comment]
Minor suggestions for the Flatpak files:
- orp.audiveris.audiveris.`desktop`: 
   - Instead of "Convert sheet music to **XML**", we should use "Convert sheet music to **MusicXML**". XML term is very broad and tells nothing about the contained information.
   - I don't understand why `lilypond` appears in the keywords.
- orp.audiveris.audiveris.`metainfo.xml`: 
   - `screenshots` element: we should provide better Audiveris images than these old ones.
   - `release` element for 5.4: we should remove this element (content is almost void and 5.4 is not released yet) and rewrite it when we are ready to release.

**Comment by mwilck** [comment]
>       * Instead of "Convert sheet music to **XML**", we should use "Convert sheet music to **MusicXML**". XML term is very broad and tells nothing about the contained information.

ok

>       * I don't understand why `lilypond` appears in the keywords.

Well ... because *my original, subjective motivation* to try out Audiveris had been to create lilypond files from printed sheet music (which requires an additional conversion step). I agree it's kind of misleading. At least other music typesetting programs like MuseScore should be mentioned, too.

I'm happy to change this to whatever you suggest.

> 
>     * orp.audiveris.audiveris.`metainfo.xml`:
>       
>       * `screenshots` element: we should provide better Audiveris images than these old ones.

Yes, and it would be wonderful if we had reliable permalinks to up-to-date screen shots on audiveris.org.

>       * `release` element for 5.4: we should remove this element (content is almost void and 5.4 is not released yet) and rewrite it when we are ready to release.

Right, probably. This is one aspect of the flatpak submission / maintenance process I don't fully understand yet. In particular I am not sure if we could submit a 5.4 release to the flathub "beta" channel before it's officially released.

**Comment by hbitteur** [comment]
> At least other music typesetting programs like MuseScore should be mentioned, too.

OK, let's mention Lilypond, Musescore and Finale

> Yes, and it would be wonderful if we had reliable permalinks to up-to-date screen shots on audiveris.org.

I will take care of finding better images.

> In particular I am not sure if we could submit a 5.4 release to the flathub "beta" channel before it's officially released.

I agree this is not very clear for me either.

In our case, we can consider that this very first submission to Flathub was a kind of beta-testing of our submission data and process.

I would suggest that the next Audiveris submission to Flathub will be tried only when the following conditions are both met:
- We have converged between your ad-hoc repositories and Audiveris repository,
- Audiveris 5.4 is officially released on Github (a matter of a couple of months, it is still in alpha status on development branch).

**Comment by mwilck** [comment]
> > Just like we did with Windows installer, we now need to automate this publishing mechanism. Certainly by defining specific tasks in the Gradle script. We need to work with @mwilck on this.

> Could we handle this (the process described in Martin's README file) via the `build.gradle` file hosted in Audiveris GitHub project? 

My plan is actually to try to merge my https://github.com/mwilck/Audiveris-flatpak stuff into the Audiveris code base by creating a PR moving all of it into the `dev/flatpak` subdirectory. Perhaps after fixing some issues that have come up in the flathub review.

Integrating that into the gradle workflow would be the last step for me, because I (being a Linux person and not a Java person) have very little knowledge about gradle. So, help with gradle would be highly appreciated :grin:

> This is what we did for Windows installer, and for the schemas documentation as well (using a dedicated sub-project). Problem: here we need to use tools like `flatpak-builder` which, to my knowledge, need to be run on Linux.

Right. I doubt flatpak-builder will ever run on Windows. But it might be possible to automate this using GitHub workflows, where using a Linux worker is no big issue. That should be the ultimate goal, once we have the manual flatpak creation working.

The current flatpak manifest for audiveris is generated from several parts. There is the (mostly static) `.yml.in` file, the list of tesseract languages to include, and most importantly, the list of dependencies, which is generated by a script. In the flathub review, I was asked to move this list auto-generated sources into a separate file. I eventually understood how this works, this is the next step I intend to do. It's just a minor change to my scripting.

Note that the "static" part isn't completely static; for example the flatpak runtime version needs to be updated from time to time. Also, as @hbitteur already noted, some information in the appstream data (`.metainfo.xml`), in particular the changelog, needs to be updated as well when a new release is attempted.

As you have seen, the current process runs the normal gradle build and collects the `.jar` files and the meta data that gradle has downloaded, creating a source list for flatpak. Because flatpak will download all these sources into a "flat" :grin: directory, we also need the `creategradlrepo.sh` script that will re-arrange them in a format the offline gradle build used by `flatpak-builder` can use. I bet my left foot that someone could come up with a gradle recipe to create these two files directly, without this clumsy "run-gradle-and-examine-what-it-did" procedure. (The python script used [this stackoverflow response](https://stackoverflow.com/questions/28436473/build-gradle-repository-for-offline-development) as a starting point).

**Comment by mwilck** [comment]
> I would suggest that the next Audiveris submission to Flathub will be tried only when the following conditions are both met:
> 
>     * We have converged between your ad-hoc repositories and Audiveris repository,
> 
>     * Audiveris 5.4 is officially released on Github (a matter of a couple of months, it is still in alpha status on development branch).

Sounds like a good plan. I am not sure about the full automation yet, but it's at least a worthwhile goal.

**Comment by hbitteur** [comment]
> Right. I doubt flatpak-builder will ever run on Windows. But it might be possible to automate this using GitHub workflows, where using a Linux worker is no big issue.

Very good idea!

> Note that the "static" part isn't completely static; for example the flatpak runtime version needs to be updated from time to time.

Yep. I noticed that flathub supports Java LTS versions 8, 11 and 17. But not Java 21 yet, even though it's the latest LTS Java version today.
I had the intention to move to Java 21 very soon, at least  on the development branch. We may have to wait.

**Comment by hbitteur** [comment]
> I don't think Finale supports Linux

I'm afraid you are right. I just checked on the Internet.

**Comment by hbitteur** [comment]
@mwilck 
Hi Martin,
I just redesigned the Gradle sub-project that builds the Windows installer.
I also did something similar to automate the generation of Audiveris handbook as a PDF.

Now I feel brave enough to start a new Gradle sub-project dedicated to your Linux installer.
I will certainly need your help here and there, but let's go! :-)
/Hervé

**Comment by hbitteur** [comment]
@mwilck 
I would like to get all your items from [ https://github.com/mwilck/Audiveris-flatpak ](https://github.com/mwilck/Audiveris-flatpak) into a dedicated folder within Audiveris Linux installer sub-project.

I plan to use <Audiveris>/linux-installer to host the needed `build.gradle` file
And your material (perhaps modified for Gradle use) would be put into <Audiveris>/linux-installer/dev folder

How could we do this? Should I manually copy every item. Or is there a more official way to proceed (I'd like to keep the link to your work, so that you get the credit).

**Comment by mwilck** [comment]
@hbitteur, 

sorry I missed these updates because I was sick and there is some confusion with email delivery from github to my various email accounts. 

I suppose you've seen my PRs #704 and #705 by now. The `dummy.yml` stuff is history now, as you may have seen. 

> languages.sh uses several "curl -s ..." which fail for some security reason 

That's strange. Were you using some proxy or the like? Or maybe your Linux installation was missing the standard set of trusted root certificates? The package is typically called `ca-certificates`.

**Comment by hbitteur** [comment]
@mwilck 
Sorry for this late answer. I had to address a couple of bug-fixes.
I'm back to the flatpak/flathub topic today.

I have pulled your PR 704 in a local folder of mine to study its content before merging it into the development branch.
In particular, I'm trying to understand what you are doing and how this could be more of less automated in Gradle.

I have to say I'd never heard of git submodules before I read your README.md recipy file. So, please be kind with my naive remarks.

Once pulled, to initialize the submodule, the recipy says we have to run:
> git submodule init --all

The `--all` option is not recognized by git. So, I removed it.

I expected to see git-managed items in folders like "dev" and then the resulting (git-ignored) artefacts written somewhere under the "build" folder. This does not seem to be the case.
For instance, the dev/flathub folder seems to gather files like the manifest `org.audiveris.audiveris.yml` file, or the appstream meta data  `org.audiveris.audiveris.metainfo.xml` file with other files that are auto-generated (`lang_sources.yml`, `dependencies.yml` and `mkgradlerepo.sh`).
I checked the `.gitignore` file in the same dev/flathub folder (submodule?). It does not ignore the auto-generated files. Why?
In short, I think we should more clearly separate the input files and the output files. Only the input files should be managed by git, and all the output files should be "cleanable" by gradle.

Focusing on the input files of Audiveris project as a whole (that is not only the dev/flathub portion),I think we should avoid any information duplication between them.
Here are some obvious examples (as of this writing):
- The project version (5.4-alpha)
- The needed Java version (17)
- Tesseract version (5....)
- etc

These constants should be defined only once, typically as symbolic constants in the `build.gradle` file.
Then any other input file should refer to the symbol name, not to its value.

For instance, "openjdk17" appears several times in `org.audiveris.audiveris.yml` file. Is there a way in such a YAML file to include symbolic names? I don't know yet.

**Comment by mwilck** [comment]
> I have to say I'd never heard of git submodules before I read your README.md recipy file. So, please be kind with my naive remarks.

I know, git submodules are somewhat difficult to get used to. But when you have overcome the first parts of the learning curve, they can be very useful. If you're interested, have a look at the chapter in the [git book](https://git-scm.com/book/en/v2/Git-Tools-Submodules).

I understand very well that if you are uncomfortable with git submodules, you don't want to start messing with them. But' I'd like to ask you to give them a fair chance :smile: You can work just normally with the audiveris repos without checking out the submodule, as long as you don't work on the flatpak. 

> > git submodule init --all
> 
> The `--all` option is not recognized by git. So, I removed it.

My bad, sorry. It's just `git submodule init; git submodule update`. `--all` is just for `git submodule deinit`, which removes the contents of the submodule directory.

> I checked the `.gitignore` file in the same dev/flathub folder (submodule?). It does not ignore the auto-generated files. Why? 

Because **for flathub**, these files are part of the submission. We **must** have them under version control (a flathub PR is nothing but a github PR, so all files that are part of the flathub submission must be in git). It doesn't matter that they are autogenerated from the audiveris code.

IOW, these file are ignored in the audiveris git repo, but not in the flathub repo. git is smart enough to understand that and not show them when you work in the main repo.

> In short, I think we should more clearly separate the input files and the output files. Only the input files should be managed by git, and all the output files should be "cleanable" by gradle.

I can do it like this if you prefer. We could have `dev/flathub` (or some other path that you might want) simply as an output directory under audiveris version control, git-ignored. But it has disadvantages compared with the submodule approach. 

1. The autogenerated files need to be copied to a checked-out flathub/audiveris repository (which wouldn't be the same in every work environment) and added and commited there before pushing / submitting to flathub. By using the submodule, the output path is fixed, and the copying is unnecessary. The autogenerated files are placed into the submodule directory directly, and a quick `git -C dev/flathub status` shows us if anything has changed and needs to be committed.
2. It's less obvious how to handle the non-autogenerated flathub files like `org.audiveris.audiveris.yml` and `org.audiveris.audiveris.metainfo.xml`. They are only strictly required for flathub, but it would be kind of awkward not to have them in the audiveris repo. I guess with the discussion below about using variable substitution, this argument is resolved; we'd keep the template in the audiveris repo and the generated output in the flathub repo.
 
Anyway, maybe we should move this discussion into #704 ?

> ... These constants should be defined only once, typically as symbolic constants in the `build.gradle` file. Then any other input file should refer to the symbol name, not to its value.

Yes of course. Would it be a problem for gradle to simply update the files in the `dev/flathub` submodule directory, using some sort of text substitution?

> For instance, "openjdk17" appears several times in `org.audiveris.audiveris.yml` file. Is there a way in such a YAML file to include symbolic names? I don't know yet.

In this specific case, I think the answer is "no". YAML-based languages support variables, like GitHub actions with it's `${{ variable }}` syntax, but it isn't standardized, and YAML itself doesn't support it. Flatpak's YAML is just plain YAML, unless I am mistaken.

We need to create a template and generate the `.yml` file from it. As a Unix guy, I'd think of some **sed** script, but I guess there are equivalent and probably more powerful means available in the gradle ecosystem.

**Comment by hbitteur** [comment]
Believe me, I had read the documentation about git submodules in [git book](https://git-scm.com/book/en/v2/Git-Tools-Submodules). A really long and impressive page!
Since you have the experience of a git submodule, and of its decoupling from the parent module, I trust you: let's use your approach to handle the flatpak/flathub work.

My contribution will be limited to the integration with gradle. The goal is to avoid as much as possible the duplication of information (because of the risk to miss some update).
I agree that. for lack of official support of user variables, the idea of template files seems the way to go.
More on this later.

**Comment by hbitteur** [comment]
If I understand correctly the various files in dev/flatpak and dev/flathub,
the flatpak manifest (**org.audiveris.audiveris.yml**) would need variables instead of hard-coded constants for:
  1. 5.4-alpha (plus its commit ID -- where do we get this ID?)
  2. openjdk17
  3. gradle-7.3-all.zip (plus its sha256 -- where do we get it?)
  4. What is not clear for me is the bunch of icon-xx.png files

The appstream meta data (**org.audiveris.audiveris.metainfo.xml**) contains one XML element per release. I suppose it was somehow copied from the Audiveris releases page on github. Manually? if so, it would not be a big deal to add a few lines for any new release.

I can't see anything in **dev/flatpak/languages.sh** and **dev/flatpak/create-flatpak-dependencies.py** files that should change from one release to the other. And both can be launched directly from a Gradle script.

So I think that, in the non-autogenerated files, only **org.audiveris.audiveris.yml** would need some template mechanism.
Gradle is based on Groovy which already provides some template mechanism. To be investigated.

**Comment by mwilck** [comment]
> If I understand correctly the various files in dev/flatpak and dev/flathub, the flatpak manifest (**org.audiveris.audiveris.yml**) would need variables instead of hard-coded constants for:
> 
>     1. 5.4-alpha (plus its commit ID -- where do we get this ID?)

I made a hackish attempt in my [flatpak workflow](https://github.com/mwilck/audiveris/blob/611195213ce1b15c27810ddbc2b69252de10f689/.github/workflows/flatpak.yml#L31) in #705. On GitHub, it's `${{ github.sha }}`, in general we can use the output of `git rev-list -n1 HEAD`.
 
>     2. openjdk17

It would make sense to stick with this for the time being, IMO. Flathub's `org.freedesktop.Sdk.Extension.openjdk` package is openjdk 21 now, so we could try building against that. But if we do this, the openjdk version might silently change under us when FlatHub updates the package. In addition, FlatHub has `org.freedesktop.Sdk.Extension.openjdk17` etc. for the LTS versions of OpenJDK:

```
OpenJDK SDK Extension	org.freedesktop.Sdk.Extension.openjdk	jdk-21.0.1+12	23.08
OpenJDK 11 SDK Extension	org.freedesktop.Sdk.Extension.openjdk11		23.08
OpenJDK 17 SDK Extension	org.freedesktop.Sdk.Extension.openjdk17		23.08
OpenJdk 8 Sdk extension	org.freedesktop.Sdk.Extension.openjdk8		23.08
```

**Note:** we may also want to have the freedesktop runtime version (here, 23.08) in the template.
 
>     3. gradle-7.3-all.zip (plus its sha256 -- where do we get it?)

Simply using sha256sum I think. It might be available from the gradle download site as well (but AFAICS maven only supplies SHA1 hashes at this time).
 
>     4. What is not clear for me is the bunch of icon-xx.png files

Flatpak needs the application icon in certain sizes, that's all. Unless you remove some of the existing icons, you don't need to bother with this part. If we had an SVG icon, we can reduce the number of icons ([see flathub requirements](https://github.com/mwilck/audiveris/blob/611195213ce1b15c27810ddbc2b69252de10f689/.github/workflows/flatpak.yml#L31)).


> The appstream meta data (**org.audiveris.audiveris.metainfo.xml**) contains one XML element per release. I suppose it was somehow copied from the Audiveris releases page on github. Manually? if so, it would not be a big deal to add a few lines for any new release.

Yes, I did this manually so far.
 
> I can't see anything in **dev/flatpak/languages.sh** and **dev/flatpak/create-flatpak-dependencies.py** files that should change from one release to the other. And both can be launched directly from a Gradle script.

I suppose both scripts will be slowly evolving over time. Note that languages.sh has an interactive mode. Btw, we may want to discuss which languages we want to ship by default – the current selection reflects my personal preferences. I suppose the selection should be the same as in the Win/Mac versions.

> So I think that, in the non-autogenerated files, only **org.audiveris.audiveris.yml** would need some template mechanism. Gradle is based on Groovy which already provides some template mechanism. To be investigated.

Great!

IMO the end goal would be to replace `create-flatpak-dependencies.py` by some gradle-internal functionality. After all, the script simply scans the result of a `gradle   build` run and gathers the dependencies that gradle has downloaded. It's certainly possible to simply make gradle print these dependencies and their sha1 hashes in some way, without acutally downloading them and running the build. If we can do that, it'd just be a matter of reformatting that output two times, for the `dependencies.yml` and the `mkgradlerepo.sh` output files.

**Comment by hbitteur** [comment]
Hi Martin,

Before we get into the improvements we are discussing here (variables, signatures, java 21, ...) I thought it was time to merge your two PRs (704 and 705) into the official Audiveris development branch.
We could then proceed to small improvements on this starting base.

A few days ago I had already pulled your 704, using the commands indicated on the related PR page:
> git checkout -b mwilck-flatpak-integration development
> git pull https://github.com/mwilck/audiveris.git flatpak-integration

I was thus able to review your files (in the 'mwilck-flatpak-integration' branch)  and start this discussion.

This morning I decided to merge this PR, using the rest of the indicated commands as follows:
```bash
$ git checkout development
warning: unable to rmdir 'dev/flathub': Directory not empty           <== WARNING HERE ========
Switched to branch 'development'
Your branch is up to date with 'origin/development'.

>$ git merge --no-ff mwilck-flatpak-integration
Merge made by the 'ort' strategy.
 .gitignore                                       |   4 +
 .gitmodules                                      |   3 +
 dev/flathub                                      |   1 +
 dev/flatpak/README.md                            | 129 ++++++++++++++++++++++++++++++
 dev/flatpak/create-flatpak-dependencies.py       | 150 +++++++++++++++++++++++++++++++++++
 dev/flatpak/languages.sh                         | 182 +++++++++++++++++++++++++++++++++++++++++++
 dev/flatpak/org.audiveris.audiveris.desktop      |  10 ---
 dev/flatpak/org.audiveris.audiveris.metainfo.xml | 144 ----------------------------------
 8 files changed, 469 insertions(+), 154 deletions(-)
 create mode 100644 .gitmodules
 create mode 160000 dev/flathub
 create mode 100644 dev/flatpak/README.md
 create mode 100755 dev/flatpak/create-flatpak-dependencies.py
 create mode 100755 dev/flatpak/languages.sh
 delete mode 100644 dev/flatpak/org.audiveris.audiveris.desktop
 delete mode 100644 dev/flatpak/org.audiveris.audiveris.metainfo.xml

$ git status
On branch development
Your branch is ahead of 'origin/development' by 6 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean

$ git pull
remote: Enumerating objects: 214, done.
remote: Counting objects: 100% (214/214), done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 214 (delta 160), reused 214 (delta 160), pack-reused 0
Receiving objects: 100% (214/214), 62.75 KiB | 845.00 KiB/s, done.
Resolving deltas: 100% (160/160), completed with 91 local objects.
From https://github.com/Audiveris/audiveris
   76a5f80f5..b7e02c25f  development -> origin/development
warning: unable to rmdir 'dev/flathub': Directory not empty          <== SAME WARNING HERE ========
Successfully rebased and updated refs/heads/development.

$ git status
On branch development
Your branch is ahead of 'origin/development' by 5 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean

$ git push origin development
Enumerating objects: 29, done.
Counting objects: 100% (29/29), done.
Delta compression using up to 8 threads
Compressing objects: 100% (25/25), done.
Writing objects: 100% (25/25), 8.92 KiB | 761.00 KiB/s, done.
Total 25 (delta 13), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (13/13), completed with 3 local objects.
To https://github.com/Audiveris/audiveris.git
   b7e02c25f..8280b604b  development -> development

$ git status
On branch development
Your branch is up to date with 'origin/development'.

nothing to commit, working tree clean
```

I think everything went fine, except the 2 warning lines about the `dev/flathub` directory not being empty.
Perhaps because this directory is handled as a specific git 'submodule'? I don't know

Anyway, if we look at the PR #704 page, we can now read that:
> This branch cannot be rebased due to conflicts

Which items are in conflict?
If I open this project in my IDE, no conflict is shown.
How can I safely complete this PR integration?

**Comment by hbitteur** [comment]
Also on PR #705 we can read:

>This branch has conflicts that must be resolved
 to resolve conflicts before continuing.
Conflicting files
dev/flathub

**Comment by hbitteur** [comment]
> > ```
> > 2. openjdk17
> > ```
> 
> 
> It would make sense to stick with this for the time being, IMO. Flathub's `org.freedesktop.Sdk.Extension.openjdk` package is openjdk 21 now, so we could try building against that. But if we do this, the openjdk version might silently change under us when FlatHub updates the package. In addition, FlatHub has `org.freedesktop.Sdk.Extension.openjdk17` etc. for the LTS versions of OpenJDK:
> 
> 
> ```
> OpenJDK SDK Extension	org.freedesktop.Sdk.Extension.openjdk	jdk-21.0.1+12	23.08
> OpenJDK 11 SDK Extension	org.freedesktop.Sdk.Extension.openjdk11		23.08
> OpenJDK 17 SDK Extension	org.freedesktop.Sdk.Extension.openjdk17		23.08
> OpenJdk 8 Sdk extension	org.freedesktop.Sdk.Extension.openjdk8		23.08
> ```
> 
> 
>     
>   
> 
> The Java version to use will be your decision. Not sure how you deal with this problem on Win/Mac – you will always have to make sure that the JRE matches what you built against, right? AFAICS, if you want to avoid nasty surprises, you have to stick with LTS versions for flatpak, meaning that this value will change vary rarely. The [flathub readme](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk) for the non-versioined openjdk package says "_OpenJDK 21 is the current latest version. This is not a long-term support (LTS) version and will be periodically updated as new JDKs are released."_.

According to [Oracle itself](https://www.oracle.com/java/technologies/java-se-support-roadmap.html#:~:text=For%20product%20releases%20after%20Java,Java%2025%20in%20September%202025.), **Java 21 is a LTS version**.
So I think we should now jump on Java 21, at least for the "development" branch, so that the next release be based on it.

**Comment by hbitteur** [comment]
> > I can't see anything in **dev/flatpak/languages.sh** and **dev/flatpak/create-flatpak-dependencies.py** files that should change from one release to the other. And both can be launched directly from a Gradle script.
> 
> I suppose both scripts will be slowly evolving over time. Note that languages.sh has an interactive mode. Btw, we may want to discuss which languages we want to ship by default – the current selection reflects my personal preferences. I suppose the selection should be the same as in the Win/Mac versions.

The predefined languages for the Windows installer are now copied from the directory `windows-installer/dev/tessdata`.
As of this writing, this means:
- deu.traineddata
- eng.traineddata
- fra.traineddata
- ita.traineddata

This should be enough for a start.
I have on the TODO list the ability for the Audiveris end-user to dynamically download additional languages directly from the Audiveris GUI.

**Comment by mwilck** [comment]
@hbitteur

> ```shell
> $ git checkout development
> warning: unable to rmdir 'dev/flathub': Directory not empty           <== WARNING HERE ========
> ```

You currently have a checked-out submodule in `dev/flathub`. In order to avoid the warning, you need to run `git submodule deinit` first.

```shell
$ git submodule deinit dev/flathub
$ git switch development
$ git merge --no-ff mwilck/flatpak-integration
$ git submodule status
-38977b6adb819de3becb8f54e36f78151278a005 dev/flathub
```
At this point, `dev/flathub` is still empty. To populate it, you need to run
```shell
$ git submodule init
Submodule 'flathub' (https://github.com/flathub/org.audiveris.Audiveris) registered for path 'dev/flathub'
$ git submodule update
Submodule path 'dev/flathub': checked out '38977b6adb819de3becb8f54e36f78151278a005'

```
Now if you want to merge #705, you have two options. You can do with or without the submodule checked out. If you merge while the submodule is checked out, you need to run `git submodule update` after the merge. This is because the flathub code has changed in #705. Your local code under `dev/flathub` is unchanged by the merge though, thus `dev/flathub` is shown as "modified", because it differs from the version registered in your `HEAD` commit. You could commit `dev/flathub` now, reverting the changes made by the merge (*not recommended!*). By running `git submodule update`, you bring the submodule in a consistent state with your `HEAD` commit.

```shell
$ git merge --no-ff -m 'merged flatpak workflow from mwilck' origin/flatpak-workflow 
Merge made by the 'ort' strategy.
 .github/workflows/flatpak.yml | 80 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 dev/flathub                   |  2 +-
 2 files changed, 81 insertions(+), 1 deletion(-)
 create mode 100644 .github/workflows/flatpak.yml
$ git status -s
 M dev/flathub
$ git submodule status
+38977b6adb819de3becb8f54e36f78151278a005 dev/flathub (remotes/origin/beta)
$ git submodule update
Submodule path 'dev/flathub': checked out '4219985d4c3bf86b9f4000fd69ee5b430fd515fd'
```

Merging might be less confusing if the submodule is not checked out:
```shell
$ git submodule deinit dev/flathub
$ git merge --no-ff -m 'merged flatpak workflow from mwilck'  mwilck/flatpak-workflow
Merge made by the 'ort' strategy.
 .github/workflows/flatpak.yml | 80 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 dev/flathub                   |  2 +-
 2 files changed, 81 insertions(+), 1 deletion(-)
 create mode 100644 .github/workflows/flatpak.yml
$ git status -s        # no output
$ git submodule status 
$ git submodule status
-4219985d4c3bf86b9f4000fd69ee5b430fd515fd dev/flathub
```
Obviously now you need to run `git submodule init; git submodule update` again if you want to *use* the `dev/flathub` stuff.

**Comment by mwilck** [comment]
> According to [Oracle itself](https://www.oracle.com/java/technologies/java-se-support-roadmap.html#:~:text=For%20product%20releases%20after%20Java,Java%2025%20in%20September%202025.), **Java 21 is a LTS version**. So I think we should now jump on Java 21, at least for the "development" branch, so that the next release be based on it.

There's actually [a flathub issue](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk/issues/38) about this, only 2 weeks old.

Probably we'll see an `openjdk21` fork on FlatHub soon. If you want / need to migrate to Java 21, we can use `openjdk` for the time being, and switch to the new LTS release when it's available.

**Comment by mwilck** [comment]
>     * deu.traineddata
>     * eng.traineddata
>     * fra.traineddata
>     * ita.traineddata
> 
> 
> This should be enough for a start.

Off-topic: I was using almost the same, but I had added latin, and `de-frak` which I hoped would be able to recognize German Gothic letters[^lang]. The result for the Gothic letters was pretty distrastrous though, I ended up removing the text part completely. :smile: 
[^lang]: I am interested in scanning old choral sheet music.

**Comment by hbitteur** [comment]
I am afraid I'm trying to run beyond my competencies about git.
The problem is that I finally pushed the 704 merge to github
See my commands above, which ended with:
```bash
$ git push origin development
Enumerating objects: 29, done.
Counting objects: 100% (29/29), done.
Delta compression using up to 8 threads
Compressing objects: 100% (25/25), done.
Writing objects: 100% (25/25), 8.92 KiB | 761.00 KiB/s, done.
Total 25 (delta 13), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (13/13), completed with 3 local objects.
To https://github.com/Audiveris/audiveris.git
   b7e02c25f..8280b604b  development -> development

$ git status
On branch development
Your branch is up to date with 'origin/development'.

nothing to commit, working tree clean
```

Now, I would like to make sure that Audiveris project is still OK on github.
For information, my local `dev/flathub` contains:
```bash
$ ls -l dev/flathub
total 123
-rwxrw-r--+ 1 herve herve   983 Dec  9 20:47 0001-add-local-dependencies-mirror.patch
-rwxrw-r--+ 1 herve herve    63 Dec  9 20:47 add-tessdata-prefix.sed
-rwxrw-r--+ 1 herve herve 51435 Dec  9 20:47 dependencies.yml
-rwxrw-r--+ 1 herve herve    33 Dec  9 20:47 flathub.json
-rwxrw-r--+ 1 herve herve   935 Dec  9 20:47 lang_sources.yml
-rwxrw-r--+ 1 herve herve 46540 Dec  9 20:47 mkgradlerepo.sh
-rwxrw-r--+ 1 herve herve   289 Dec  9 20:47 org.audiveris.audiveris.desktop
-rwxrw-r--+ 1 herve herve  4986 Dec  9 20:47 org.audiveris.audiveris.metainfo.xml
-rwxrw-r--+ 1 herve herve  3196 Dec  9 20:47 org.audiveris.audiveris.yml
```
What should I do precisely?
I suppose I don't need to do again something like:
> $ git merge --no-ff -m 'merged flatpak workflow from mwilck' origin/flatpak-workflow

**Comment by mwilck** [comment]
The development branch looks good to me. You successfully merged #704 into it.
What exactly is your concern?

I suggest you clone the GH repo into a new directory without doing anything to the submodule, and run your usual build workflow to make sure everything works as usual.

**Comment by hbitteur** [comment]
I just cloned the repo into a new directory of mine. Went fine.
Can I now integrate PR #705 on top of it?

**Comment by mwilck** [comment]
> I just cloned the repo into a new directory of mine. Went fine. Can I now integrate PR #705 on top of it?

I guess so. Note that it enables aarch64.

**Comment by hbitteur** [comment]
OK. Let's go!

**Comment by mwilck** [comment]
I just made a fresh clone of development (before #705) and ran `ev/flatpak/create-flatpak-dependencies.py` successfully. As expected, the submodule state was unchanged after running the script.

**Comment by hbitteur** [comment]
FYI, the current view on PR #705 ( look by yourself) exhibits similar warnings:

![image](https://github.com/Audiveris/audiveris/assets/22053634/47640625-db44-4e4e-abc6-414521b01304)

**Comment by mwilck** [comment]
I just verified that you can run `dev/flatpak/create-flatpak-dependencies.py` even if you don't have the `dev/flathub` submodule checked out. It works nicely. The only problem is that if you later initialize the submodule, the results of the run might be overwritten. Therefore checkout out the submodule before running the script is recommended.

You should actually be able to run `dev/flatpak/create-flatpak-dependencies.py` Windows if you have python (although there may be detail issues as I never tested it).

I'll double-check #705.

**Comment by mwilck** [comment]
The merge is actually non-trivial, because `development` is not an ancestor of my `flatpak-workflow` branch.


To do it manually, you **must** have the submodule checked-out before (well maybe there's a trick to make `git submodule update` work during a merge with conflicts, but I don't know it).

```shell
$ git merge origin/flatpak-workflow
Failed to merge submodule dev/flathub (no merge base)
CONFLICT (submodule): Merge conflict in dev/flathub
Recursive merging with submodules currently only supports trivial cases.
Please manually handle the merging of each conflicted submodule.
This can be accomplished with the following steps:
 - go to submodule (dev/flathub), and either merge commit 4219985
   or update to an existing commit which has merged those changes
 - come back to superproject and run:

      git add dev/flathub

   to record the above merge or update
 - resolve any other conflicts in the superproject
 - commit the resulting index in the superproject
Automatic merge failed; fix conflicts and then commit the result.
## 421998 is the HEAD of the flathub repo in #705 
$ git -C dev/flathub/ reset --hard  421998
## the following resolves the conflict
$ git add dev/flathub/
$ git merge --continue
```

But **don't do this now**, I'll rebase my branch; that's easier for both of us.

**Comment by mwilck** [comment]
Rebase done, you should be able to do a clean merge now.

**Comment by mwilck** [comment]
Don't worry, you won't face this kind of difficulty regularly in the future. The problem here was that you had made unrelated changes before merging #704. Therefore the result of the merge was not an ancestor of #705, and thus, for git, they were two independent branches that had both added the `dev/flathub` submodule, with different `HEAD`s.

**Comment by hbitteur** [comment]
I just merged your PR #705 and pushed. 
The page on #705 looks great now.

However page on PR #704 still has a warning.
Could you have a look?

**Comment by hbitteur** [comment]
I tried to run  `dev/flatpak/create-flatpak-dependencies.py`.
I modified just one line (number 82) from

> subprocess.call(["./gradlew", "-q", "-g", temp_home, "build"])

to: 
> subprocess.call(["cmd", "/c", "gradlew.bat", "-q", "-g", temp_home, "build"])

to be able to run in under Windows.
It seems to went fine (building the project and its dependencies).
In `dev/flathub`, as expected, two files were modified: `dependencies.yml` and `mkgradlerepo.sh`
This sound good.

A diff between old and new files gave interesting results.  between old and new.
Here are the differences for dependencies.yml.
Missing in new:
        url: https://repo1.maven.org/maven2/com/sun/activation/all/1.2.0/all-1.2.0.pom
        url: https://repo1.maven.org/maven2/commons-logging/commons-logging/1.2/commons-logging-1.2.jar
        url: https://repo1.maven.org/maven2/commons-logging/commons-logging/1.2/commons-logging-1.2.pom
        url: https://repo1.maven.org/maven2/net/java/jvnet-parent/1/jvnet-parent-1.pom
        url: https://repo1.maven.org/maven2/net/java/jvnet-parent/5/jvnet-parent-5.pom
        url: https://repo1.maven.org/maven2/net/jcip/jcip-annotations/1.0/jcip-annotations-1.0.jar
        url: https://repo1.maven.org/maven2/net/jcip/jcip-annotations/1.0/jcip-annotations-1.0.pom
        url: https://repo1.maven.org/maven2/org/apache/apache/10/apache-10.pom
        url: https://repo1.maven.org/maven2/org/apache/apache/19/apache-19.pom
        url: https://repo1.maven.org/maven2/org/apache/apache/21/apache-21.pom
        url: https://repo1.maven.org/maven2/org/apache/apache/23/apache-23.pom
        url: https://repo1.maven.org/maven2/org/apache/commons/commons-parent/34/commons-parent-34.pom
        url: https://repo1.maven.org/maven2/org/apache/commons/commons-parent/48/commons-parent-48.pom
        url: https://repo1.maven.org/maven2/org/apache/commons/commons-parent/52/commons-parent-52.pom
        url: https://repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar
        url: https://repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.pom
        url: https://repo1.maven.org/maven2/org/hamcrest/hamcrest-parent/1.3/hamcrest-parent-1.3.pom
        url: https://repo1.maven.org/maven2/org/sonatype/oss/oss-parent/7/oss-parent-7.pom
Why have these dependencies disappeared between old(Linux) and new (Windows)?

We have normal differences between old (Linux) and new (Windows) versions of leptonica and tesseract files:
- linux-arm64 and linux-x86_64 (in old)
- windows-x86_64 (in new)

**Comment by mwilck** [comment]
> However page on PR #704 still has a warning. Could you have a look?

I don't know. I just closed it. Let's observe if this pattern repeats. I don't expect that it will.

**Comment by mwilck** [comment]
> I have in mind to configure Audiveris git repository so that its _default_ branch is set to "development" rather than "master". See this related [git documentation](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch). What do you think?

Sure, why not. It's up to you. It depends on what you want people to use primarily. If "development" is stable enough for those who use git, make it the default branch. A word of warning: This may have the side effect that the "master" branch may become stale sooner or later.

**Comment by mwilck** [comment]
> I tried to run `dev/flatpak/create-flatpak-dependencies.py`. I modified just one line (number 82) from
> 
> > subprocess.call(["./gradlew", "-q", "-g", temp_home, "build"])
> 
> to:
> 
> > subprocess.call(["cmd", "/c", "gradlew.bat", "-q", "-g", temp_home, "build"])
> 
> to be able to run in under Windows. 

`cmd /c` won't work under Linux, but we could make this call conditional on `sys.platform`.

> Why have these dependencies disappeared between old(Linux) and new (Windows)?

Honestly I have no idea. Can the dependencies be platform-specific? Can you check if these files have been downloaded under `dev/flatpak/gradle` ?

**Comment by hbitteur** [comment]
> Honestly I have no idea. Can the dependencies be platform-specific? Can you check if these files have been downloaded under `dev/flatpak/gradle` ?

I checked under `dev/flatpak/gradle/caches`, none of them has been downloaded.
After all, this is consistent with the fact that they don't appear in the resulting `dependencies.yml` file.

Perhaps you should check the cache on your (Linux) side.
Do these items exist there?

**Comment by mwilck** [comment]
FTR: [The Java 21 LTS Flatpak has now been added to Flathub](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk21/).

**Comment by hbitteur** [comment]
Good to know that.
Thanks

**Comment by hbitteur** [comment]
@scohenjanes5 
Martin Wilck did an excellent job making Audiveris available on Flathub.

I personally struggled for a few years to develop and deploy the Flatpak version, without much success, perhaps because my Linux knowledge is quite limited.

Now that we can release Audiveris on Windows, Linux, and macOS in a flash thanks to the Github workflow, I admit that my motivation for Flatpak has faded.

**Comment by scohenjanes5** [comment]
To make the case for the Flatpak: Some Linux users cannot use the `.deb` package at all since some flavors of Linux use different packaging formats. Flatpaks are distro-agnostic, and they are in a central repository that every "software center" can automatically access. This saves us from having to revisit the GitHub page and manually install every time there is an update. It also means that to service all linux users you only need to publish this one type of package rather than make `.deb` `.rpm`, etc.

I believe you can automatically trigger the workflow on `release` instead of `workflow_dispatch`, the latter of which I think requires someone to push a button. This way you can "forget" about it and the flatpak will be made automatically upon release.

See the [workflow docs](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#release) for something like this:
```yaml
on:
  release:
    types: [published]
```

**Comment by mwilck** [comment]
Both @hbitteur and me understand the importance of flatpak. Indeed the audiveris build procedure produces a (local) flatpak package. We just have a few small hickups with publishing on Flathub, where additional checks are applied that complicate matters.
